server/src/lib/model/statistics.ts

```diff
@@ -0,0 +1,239 @@
+import lazyCache from '../lazy-cache';
+import { getMySQLInstance } from './db/mysql';
+import { TableNames, TimeUnits } from 'common';
+const db = getMySQLInstance();
+
+type StatisticsCount = {
+  total_count: number;
+  date?: string;
+};
+
+const FILTERS = {
+  rejected: 'is_valid = false',
+  hasEmail: 'email IS NOT null',
+};
+
+type Filters = 'rejected' | 'hasEmail';
+
+type QueryOptions = {
+  groupByColumn?: string;
+  isDistinct?: boolean;
+  isDuplicate?: boolean;
+  filter?: Filters;
+};
+
+/**
+ * Attach metadata about the stats (like when it was fetched)
+ *
+ */
+const getResponseMetadata = () => {
+  const today = new Date();
+  return {
+    last_fetched: today,
+  };
+};
+
+const queryStatistics = async (
+  tableName: TableNames,
+  options?: QueryOptions
+) => {
+  const isDistinct = options?.isDistinct ?? false;
+  const isDuplicate = options?.isDuplicate ?? false;
+  let monthlyIncrease, totalCount;
+
+  if (isDistinct) {
+    // Queries with distinct have to use group by
+    monthlyIncrease = await getUniqueMonthlyContributions(tableName, options);
+    totalCount = await getUniqueSpeakersTotal();
+  } else if (isDuplicate) {
+    // Specific query flow since logic is more complicated
+    monthlyIncrease = await getMonthlyDuplicateSentences();
+    totalCount = await getTotalDuplicateSentences();
+  } else {
+    // Simple queries
+    monthlyIncrease = await getMonthlyContributions(tableName, options);
+    totalCount = await getTotal(tableName, options);
+  }
+  const yearlySum = monthlyIncrease.reduce(
+    (total: number, row) => (total += row.total_count),
+    0
+  );
+  totalCount = totalCount.total_count;
+
+  return { yearlySum, totalCount, monthlyIncrease };
+};
+
+const formatStatistics = async (
+  yearlySum: number,
+  totalCount: number,
+  monthlyIncrease: StatisticsCount[]
+) => {
+  //format raw query data
+  const monthlyContributions = monthlyIncrease.reduce((obj: any, row) => {
+    obj[row.date] = row.total_count;
+    return obj;
+  }, {});
+
+  let currentSum = 0;
+  const monthlyRunningTotals = monthlyIncrease.reduce((obj: any, row) => {
+    const diff = yearlySum - currentSum;
+    currentSum += row.total_count;
+    obj[row.date] = diff;
+    return obj;
+  }, {});
+
+  return {
+    yearly_sum: yearlySum,
+    total_count: totalCount,
```

Comment:
**Question/Thought**: It seems `totalCount` is not really used anywhere else in this function besides being passed in as a parameter and returned from the function. So I was wondering if it was better to return it from the `getStatistics` function (after line 219). If there is another reason for it to be passed into this function I'm fine with leaving it as it is now.

Reply:
I felt it made more sense to have one function responsible for building the response (it builds the response object and changes the key name). It'd theoretically make testing simpler (there is no testing currently however) because it removes a side effect of building in two places